controllers/state_of_the_world.go

```diff
@@ -0,0 +1,167 @@
+		controller.ManagedBy(manager),
+		controller.WithLogger(logger),
+		controller.WithClient(client),
+		controller.WithRunnable("kuadrant watcher", controller.Watch(&kuadrantv1beta1.Kuadrant{}, kuadrantv1beta1.KuadrantResource, metav1.NamespaceAll)),
+		controller.WithRunnable("gatewayclass watcher", controller.Watch(&gwapiv1.GatewayClass{}, controller.GatewayClassesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("gateway watcher", controller.Watch(&gwapiv1.Gateway{}, controller.GatewaysResource, metav1.NamespaceAll)),
+		controller.WithRunnable("httproute watcher", controller.Watch(&gwapiv1.HTTPRoute{}, controller.HTTPRoutesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("dnspolicy watcher", controller.Watch(&kuadrantv1alpha1.DNSPolicy{}, kuadrantv1alpha1.DNSPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("tlspolicy watcher", controller.Watch(&kuadrantv1alpha1.TLSPolicy{}, kuadrantv1alpha1.TLSPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("authpolicy watcher", controller.Watch(&kuadrantv1beta2.AuthPolicy{}, kuadrantv1beta2.AuthPoliciesResource, metav1.NamespaceAll)),
+		controller.WithRunnable("ratelimitpolicy watcher", controller.Watch(&kuadrantv1beta2.RateLimitPolicy{}, kuadrantv1beta2.RateLimitPoliciesResource, metav1.NamespaceAll)),
+		controller.WithPolicyKinds(
+			kuadrantv1alpha1.DNSPolicyKind,
+			kuadrantv1alpha1.TLSPolicyKind,
+			kuadrantv1beta2.AuthPolicyKind,
+			kuadrantv1beta2.RateLimitPolicyKind,
+		),
+		controller.WithObjectKinds(
+			kuadrantv1beta1.KuadrantKind,
+			ConfigMapGroupKind),
+		controller.WithObjectLinks(
+			kuadrantv1beta1.LinkKuadrantToGatewayClasses,
+		),
+		controller.WithReconcile(buildReconciler(client)),
+	}
+
+	if topologyNamespace != "" {
+		controllerOpts = append(
+			controllerOpts,
+			controller.WithRunnable(
+				"topology configmap watcher",
+				controller.Watch(
+					&corev1.ConfigMap{}, controller.ConfigMapsResource,
+					topologyNamespace,
+					controller.FilterResourcesByLabel[*corev1.ConfigMap](fmt.Sprintf("%s=true", kuadrant.TopologyAnnotation)))))
+	}
+
+	return controller.NewController(controllerOpts...)
+}
+
+func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+	reconciler := &controller.Workflow{
+		Precondition: NewEventLogger().Log,
+	}
+	if topologyNamespace != "" {
+		reconciler.Tasks = append(reconciler.Tasks, NewTopologyFileReconciler(client).Reconcile)
+	}
+
+	return reconciler.Run
+}
+
+type TopologyFileReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewTopologyFileReconciler(client *dynamic.DynamicClient) *TopologyFileReconciler {
+	return &TopologyFileReconciler{Client: client}
+}
+
+func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error) {
+	logger := controller.LoggerFromContext(ctx).WithName("topology file")
+
+	if topologyNamespace == "" {
```

Comment:
nit:

Can we please make `TopologyNamespace` a field of the `TopologyFileReconciler` struct (filled from the `topologyNamespace` package-wide variable when the reconciler is initialised)?

I'm thinking on the (de)coupling of this type from the rest, maybe so it can be moved to its own file or even package at some point.

Reply:
Should the `TopologyFileReconciler` raise an error if there is a empty string passed into to the constructor. If so, how would you suggest handling the error in the [buildReconciler](https://github.com/Kuadrant/kuadrant-operator/blob/c18da761d52b2f4bf4fdf98a43f3a0c2f6350c11/controllers/state_of_the_world.go#L75-L84) function?